extensions/emmet/src/abbreviationActions.ts

```diff
@@ -32,48 +38,134 @@ export function wrapWithAbbreviation(args: any) {
+			return { syntax, abbreviation, rangeToReplace: rangesAndContent.originalRange, textToWrap, filter };
+		});
 
-		editor.selections.forEach(selection => {
-			let rangeToReplace: vscode.Range = selection.isReversed ? new vscode.Range(selection.active, selection.anchor) : selection;
-			if (rangeToReplace.isEmpty) {
-				let { active } = selection;
-				let currentNode = getNode(rootNode, active, true);
-				if (currentNode && (currentNode.start.line === active.line || currentNode.end.line === active.line)) {
-					rangeToReplace = new vscode.Range(currentNode.start, currentNode.end);
-				} else {
-					rangeToReplace = new vscode.Range(rangeToReplace.start.line, 0, rangeToReplace.start.line, editor.document.lineAt(rangeToReplace.start.line).text.length);
-				}
-			}
+		return applyPreview(editor, expandAbbrList, rangesToReplace);
+	}
 
-			const firstLineOfSelection = editor.document.lineAt(rangeToReplace.start).text.substr(rangeToReplace.start.character);
-			const matches = firstLineOfSelection.match(/^(\s*)/);
-			const preceedingWhiteSpace = matches ? matches[1].length : 0;
+	// On inputBox closing
+	return abbreviationPromise.then(inputAbbreviation => {
+		return makeChanges(inputAbbreviation, previewMade, true);
+	});
+}
 
-			rangeToReplace = new vscode.Range(rangeToReplace.start.line, rangeToReplace.start.character + preceedingWhiteSpace, rangeToReplace.end.line, rangeToReplace.end.character);
-			let textToWrap = rangeToReplace.isSingleLine ? ['$TM_SELECTED_TEXT'] : ['\n\t$TM_SELECTED_TEXT\n'];
-			expandAbbrList.push({ syntax, abbreviation, rangeToReplace, textToWrap, filter });
-		});
+function revertPreview(editor: vscode.TextEditor, rangesToReplace: PreviewRangesWithContent[]): Thenable<any> {
+	return editor.edit(builder => {
+		for (let i = 0; i < rangesToReplace.length; i++) {
+			builder.replace(rangesToReplace[i].previewRange, rangesToReplace[i].originalContent);
+			rangesToReplace[i].previewRange = rangesToReplace[i].originalRange;
+		}
+	}, { undoStopBefore: false, undoStopAfter: false });
+}
 
-		return expandAbbreviationInRange(editor, expandAbbrList, true);
-	});
+function applyPreview(editor: vscode.TextEditor, expandAbbrList: ExpandAbbreviationInput[], rangesToReplace: PreviewRangesWithContent[]): Thenable<any> {
+	const anyExpandAbbrInput = expandAbbrList[0];
+	let expandedText = expandAbbr(anyExpandAbbrInput);
+	let linesInserted = 0;
+
+	if (expandedText) {
+		return editor.edit(builder => {
+			for (let i = 0; i < rangesToReplace.length; i++) {
+				if (i) {
+					expandedText = expandAbbr(expandAbbrList[i]);
+				}
+				let thisRange = rangesToReplace[i].previewRange;
+				let indentPrefix = '';
+				let preceedingText = editor.document.getText(new vscode.Range(new vscode.Position(thisRange.start.line, 0), thisRange.start));
+				// If there is only whitespace before the text to wrap, take that as prefix. If not, take as much whitespace as there is before text appears.
+				if (!preceedingText.match(/[^\s]/)) {
```

Comment:
Instead of this if/else, why not follow the same logic as `preceedingWhiteSpace` in https://github.com/Microsoft/vscode/blob/master/extensions/emmet/src/abbreviationActions.ts#L66-L68?